numpy/core/src/umath/ufunc_object.c

```diff
@@ -3402,38 +3758,65 @@ static int
 reduce_type_resolver(PyUFuncObject *ufunc, PyArrayObject *arr,
                         PyArray_Descr *odtype, PyArray_Descr **out_dtype)
 {
-    int i, retcode;
+    int retcode;
     PyArrayObject *op[3] = {arr, arr, NULL};
     PyArray_Descr *dtypes[3] = {NULL, NULL, NULL};
+    PyArray_Descr *op_dtypes[3] = {NULL, NULL, NULL};
     const char *ufunc_name = ufunc_get_name_cstr(ufunc);
-    PyObject *type_tup = NULL;
+
+    /* Using borrowed references */
+    op_dtypes[0] = PyArray_DESCR(arr);
+    op_dtypes[1] = op_dtypes[0];
 
     *out_dtype = NULL;
 
-    /*
-     * If odtype is specified, make a type tuple for the type
-     * resolution.
-     */
-    if (odtype != NULL) {
-        type_tup = PyTuple_Pack(3, odtype, odtype, Py_None);
-        if (type_tup == NULL) {
-            return -1;
+    /* Use the type resolution function to find our loop */
+    if (ufunc->type_resolver != NULL) {
+        /*
+         * If odtype is specified, make a type tuple for the type
+         * resolution.
+         * Behaviour is slightly strange but unused within numpy now.
+         */
+        PyObject *type_tup = NULL;
+        if (odtype != NULL) {
+            type_tup = PyTuple_Pack(3, odtype, odtype, Py_None);
+            if (type_tup == NULL) {
+                return -1;
+            }
         }
+
+        retcode = ufunc->type_resolver(
+                            ufunc, NPY_UNSAFE_CASTING,
+                            op, type_tup, dtypes);
+        Py_DECREF(type_tup);
```

Comment:
XDECREF, since type_tup can be `NULL` here

Reply:
Whoops. Need to undo the changes in type resolution and run the full test suit using this code branch as well. Probably, should try to split this all up into their own commits (or even PRs).